### PR TITLE
Cv default predict

### PR DESCRIPTION
**Pull request by tdhock**
Hi @parismita I have a suggestion for the user interface of the *.cv functions.

Typically a useR would expect to be able to do
```r
fit <- compute_model(X, y)
pred.vec <- predict(fit, Xtest)
```
in order to get a vector of predictions for a test data set. 

This can not be done using the current mmit.cv/mmif.cv functions (I coded two tests that should fail).

There are two issues that should be fixed:
1. *.cv(X,y) gives an error: param_grid is missing with no default. Can you please provide some sensible defaults?
2. predict(fit, Xtest) does not work on objects returned from *.cv functions. Can you please assign classes "mmit.cv" and "mmif.cv" to the objects returned by those functions, and define appropriate S3 methods? Full docs are here http://adv-r.had.co.nz/S3.html but essentially you just need to define two new functions, named predict.mmit.cv and predict.mmif.cv, which compute the vector of predictions, for example https://github.com/tdhock/penaltyLearning/blob/master/R/IntervalRegression.R#L710

### CODE REVIEW

**Comment by parismita** [comment]
yeah...I'll work on the second issue

regarding the param_grid, the default values are already assigned inside mmit.cv/mmif.cv, so we can assign param_grid = NULL in the argument to solve the 1st issue.

**Comment by tdhock** [comment]
ok great in that case the fixes should be easy.

**Comment by parismita** [comment]
yeah

File: Rpackage/R/mmif.cv.predict.R [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#' feature.mat <- data.frame(feature.mat)
+#' 
+#' fit <- mmif.cv(feature.mat, target.mat)
+#' 
+#' pred <- mmif.cv.predict(fit)
+#' 
+mmif.cv.predict <- function(object, newdata = NULL, perm = NULL){
+  
+  all_pred <- NULL
+  forest <- object$best_estimator
```
Comment (tdhock) [inline Rpackage/R/mmif.cv.predict.R:34]: hi @parismita  in order to get `predict(mmif.obj, newX)` working, you need to change the name of these functions. the S3 class system requires that the functions be named as method_name.class_name, so in this case it is predict.mmif and predict.mmif.cv


after you have defined predict.mmif, you can you it in the definition of predict.mmif.cv, e.g.
```r
predict.mmif.cv <- function(object, X, ...){
  predict(object$best_estimator, X, ...)
}
```

File: Rpackage/tests/testthat/test-cv.R [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+test.fold <- 1L
+is.train <- fold.vec != test.fold
+X.train <- data.frame(neuroblastomaProcessed$feature.mat[is.train,])
+y.train <- neuroblastomaProcessed$target.mat[is.train,]
+X.test <- data.frame(neuroblastomaProcessed$feature.mat[!is.train,])
+y.test <- neuroblastomaProcessed$target.mat[!is.train,]
+
+test_that("mmit.cv(X,y) works", {
+  fit <- mmit.cv(X.train, y.train)
+  pred.vec <- mmit.cv.predict(fit, X.test)
```
Comment (tdhock) [inline Rpackage/tests/testthat/test-cv.R]: @parismita please change this test back to what I proposed (using the predict S3 method is what an R user should be able to do), e.g.
```
fit <- mmit.cv(X.train, y.train)
pred.vec <- predict(fit, X.test)
```

File: Rpackage/R/mmif.cv.predict.R (hunk 1 above)
Comment (parismita) [reply Rpackage/R/mmif.cv.predict.R:34]: ok

File: Rpackage/tests/testthat/test-cv.R (hunk 2 above)
Comment (parismita) [reply Rpackage/tests/testthat/test-cv.R]: so if we define predict.mmit.cv, we can call it by directly using predict?

Comment (tdhock) [reply Rpackage/tests/testthat/test-cv.R]: right.

whenever you do predict(x) that is equivalent to predict.class_of_x(x)

Comment (tdhock) [reply Rpackage/tests/testthat/test-cv.R]: if you look at the definition of the predict function, it just calls UseMethod("predict") which handles the "method dispatch" -- it figures out what class your object is, and calls the corresponding predict.* function

Comment (parismita) [reply Rpackage/tests/testthat/test-cv.R]: @tdhock  should I do this for mmif as well and change mmif.predict to predict?

only mmit must return a party object of trees...thus we cant do "predict" directly on it.

Comment (parismita) [reply Rpackage/tests/testthat/test-cv.R]: shall i change each predict function to predict.* for uniformity, then we can define mmif return object as s3 class as well.

Comment (tdhock) [reply Rpackage/tests/testthat/test-cv.R]: right

**Comment by parismita** [comment]
@tdhock Hi, the two tasks are completed, can you review it, I need to pull the latest changes into the Adaboost PR to complete that one. Thanks

**Comment by tdhock** [comment]
hi @parismita thanks for the update. On Travis-R I see the following issues related to docs / method declarations, can you please fix?

```
* checking S3 generic/method consistency ... WARNING
predict:
  function(object, ...)
predict.mmif:
  function(forest, test_feature.mat)
predict:
  function(object, ...)
predict.mmif.cv:
  function(object, newdata)
predict:
  function(object, ...)
predict.mmit:
  function(tree, newdata, perm)
predict:
  function(object, ...)
predict.mmit.cv:
  function(object, newdata, perm)
See section ‘Generic functions and methods’ in the ‘Writing R
Extensions’ manual.
```
for these above you need to make sure that your predict functions (e.g. predict.mmit) always have the dots arguments at the end (even if they are ignored) in order to be consistent with the generic which has those dots. also for   function(forest, test_feature.mat) you should change "forest" to "object" to be consistent with the generic. for example this is what I do for a custom predict method https://github.com/tdhock/penaltyLearning/blob/master/R/IntervalRegression.R#L730 which does not generate any warnings.

```
* checking for code/documentation mismatches ... WARNING
Codoc mismatches from documentation object 'predict.mmif.cv':
predict.mmif.cv
  Code: function(object, newdata = NULL)
  Docs: function(object, newdata = NULL, perm = NULL)
  Argument names in docs not in code:
    perm
```

for these looks like you probably just need to run `devtools::document()`

**Comment by parismita** [comment]
ok...I'll check that, maybe I forgot to update the doc for the predict functions

**Comment by parismita** [comment]
@tdhock I updated the docs and solved that issue, it was because I didn't define S3method in namespace.

**Comment by tdhock** [comment]
ok looks good to me @parismita 
since I initiated the PR, somebody else needs to review before it can be merged (I think), can you please review + merge?

**Comment by parismita** [comment]
ok

**Review by parismita — APPROVED** [review]

**Pull request merged**